galley/pkg/runtime/projections/serviceentry/handler.go

```diff
@@ -0,0 +1,300 @@
+func (p *Handler) Handle(event resource.Event) {
+	if event.Kind == resource.FullSync {
+		p.distribute = true
+		p.notifyChanged()
+		return
+	}
+
+	switch event.Entry.ID.Collection {
+	case metadata.Endpoints.Collection:
+		// Update the projections
+		p.handleEndpointsEvent(event)
+	case metadata.Service.Collection:
+		// Update the projections
+		p.handleServiceEvent(event)
+	case metadata.Node.Collection:
+		// Just add the node to the cache.
+		p.nodesHandler.Handle(event)
+	case metadata.Pod.Collection:
+		// Just add the pod to the cache.
+		p.podsHandler.Handle(event)
+	default:
+		scope.Warnf("received event with unexpected collection: %v", event.Entry.ID.Collection)
+	}
+}
+
+// Builds the snapshot of the current resources.
+func (p *Handler) BuildSnapshot() snapshot.Snapshot {
+	p.mcpLock.Lock()
+	defer p.mcpLock.Unlock()
+
+	now := time.Now()
+	monitoring.RecordProcessorSnapshotPublished(p.pendingEvents, now.Sub(p.lastSnapshotTime))
+	p.lastSnapshotTime = now
+
+	b := snapshot.NewInMemoryBuilder()
+
+	entries := make([]*mcp.Resource, 0, len(p.mcpResources))
+	for _, r := range p.mcpResources {
+		entries = append(entries, r)
+	}
+	version := fmt.Sprintf("%d", p.version)
+	b.Set(collection.String(), version, entries)
+
+	p.pendingEvents = 0
+
+	return b.Build()
+}
+
+func (p *Handler) handleServiceEvent(event resource.Event) {
+	entry := event.Entry
+	fullName := entry.ID.FullName
+
+	switch event.Kind {
+	case resource.Added, resource.Updated:
+
+		// Check to see if the service's version has changed.
+		curEntry, ok := p.services[fullName]
+		if ok && curEntry.ID.Version == entry.ID.Version {
+			scope.Debugf("Received service for the current, known version: %v", event)
+			return
+		}
+
+		// Store the service.
+		p.services[fullName] = entry
+
+		// Get the associated endpoints, if available.
+		var endpoints *coreV1.Endpoints
+		if endpointsEntry, ok := p.endpoints[fullName]; ok {
+			endpoints = endpointsEntry.Item.(*coreV1.Endpoints)
+		}
+
+		// Convert to an MCP resource to be used in the snapshot.
+		mcpEntry, ok := p.toMcpResource(entry, endpoints)
+		if !ok {
+			return
+		}
+		p.setMcpEntry(fullName, mcpEntry)
+
+		p.updateVersion()
+
+	case resource.Deleted:
+		// Delete the Service and ServiceEntry
+		delete(p.services, fullName)
+		delete(p.mcpResources, fullName)
+		p.updateVersion()
+	default:
+		scope.Errorf("Unknown event kind: %v", event.Kind)
+	}
+}
+
+func (p *Handler) handleEndpointsEvent(event resource.Event) {
+	entry := event.Entry
+	fullName := entry.ID.FullName
+
+	switch event.Kind {
+	case resource.Added, resource.Updated:
+
+		// Check to see if the version has changed.
+		curEntry, ok := p.endpoints[fullName]
+		if ok && curEntry.ID.Version == entry.ID.Version {
+			scope.Debugf("Received endpoints for the current, known version: %v", event)
+			return
+		}
+
+		p.endpoints[fullName] = entry
+
+		// Look up the service associated with the endpoints.
+		svcEntry, ok := p.services[fullName]
```

Comment:
Is the name of the endpoint the same as the name of the service?

Reply:
that's my understanding.  @costinm can you confirm?

Reply:
Then how does it work with multiple endpoints?

Reply:
See my previous reply.

Reply:
yes, confirmed.

Reply:
It works with multiple endpoints because the endpoints cache is keyed by the full service name and each cache entry contains the full list of endpoints.